### DISCUSSION TITLE
OpenSSL 3.0.13 default check Operation not supported using engine

### DISCUSSION THREAD

**Discussion by vyasatavi** [openssl/openssl#25588] [Q&A]
We are trying to support Safenet Luna gemengine and OpenSSL 3.0.13 throws a cryptic error when using the reference pem key while generating CSR/cert. This was not an issue with OpenSSL 1.1.1.

sudo openssl req -engine gem -x509 -new -days 365 -key key.pem  -nodes -extensions usr_cert  -subj "/CN=rsa.com" 
Engine "gem" set.
40673A2B207B0000:error:03000093:digital envelope routines:default_check:command not supported:../crypto/evp/ctrl_params_translate.c:329:

What does that mean?

**Comment by mattcaswell** [comment]
That looks like a bug. In a debug build this would have crashed with an assertion failure. Unfortunately we don't have enough information to track this down. Is it possible for you run with a debug build of OpenSSL? A backtrace leading to this would be very helpful.

**Comment by vyasatavi** [comment]
Hi @mattcaswell , tried with 
./Configure debug no-shared
make && make install
Still did not see any crash but the same error print showed up. 

Do you expect an assert()? Do you know if I'm doing it right?
-Vyas

**Reply by nhorman** [reply]
you need to configure with the debug option enabled to get asserts

**Reply by vyasatavi** [reply]
Tried ./config -d shared as well. Is there a page explaining this?

**Reply by mattcaswell** [reply]
The "make install" command above is probably installing OpenSSL into `/usr/local` but my guess is you are still picking up the system OpenSSL when you run your command. Ensure you are using the correct version of OpenSSL (i.e. the debug build you just created and not the system version). Check your `$PATH`  (to check you are picking up the correct version of the OpenSSL binary) as well as your `$LD_LIBRARY_PATH` (to ensure you are picking up the correct versions of libcrypto/libssl - not the system versions).

Try `which openssl` to tell you where you are picking up the OpenSSL binary from and ``ldd `which openssl` `` to check where you are picking up your version of libcrypto/libssl

**Reply by vyasatavi** [reply]
@mattcaswell Thank you. OpenSSL is indeed being picked up from /usr/local/bin/openssl
<img width="544" alt="Screenshot 2024-10-14 at 11 40 12 AM" src="https://github.com/user-attachments/assets/60dbe333-0dd8-4880-9b13-7e5ce32448e8">
 Setting LD_LIBRARY_PATH to my openssl -3.0.13 build folder is still giving the same error. 
<img width="718" alt="Screenshot 2024-10-14 at 11 46 03 AM" src="https://github.com/user-attachments/assets/c4e55369-61ec-4324-91ab-6c2498c09056">

**Comment by vyasatavi** [comment]
Bump! I still could not figure out how to run with -debug option when running openssl req cmd.

**Comment by Joelone** [comment]
Another data point, OpenSSL 3.0.8 and gemengine 1.6 with the debug option enable.
There is an additional line of PKCS#11 error from gem engine, but I don't see any assert.
```
# openssl req -engine gem -new -key key.pem -out my-csr.csr -subj "/CN=rsa.com"
Engine "gem" set.
80F296D9E87F0000:error:03000093:digital envelope routines:default_check:command not supported:crypto/evp/ctrl_params_translate.c:329:
Cannot write random bytes:
80F296D9E87F0000:error:4000008D:Luna gem engine:ERR_LUNACA3_error:PKCS#11 error:e_gem.c:1550:
```
**Update**
The previous error was due to a login issue. After resolving that, here is the new error message:
```
# openssl req -engine gem -new -key key.pem -out my-csr.csr -subj "/CN=rsa.com"
Engine "gem" set.
80524C50DB7F0000:error:03000093:digital envelope routines:default_check:command not supported:crypto/evp/ctrl_params_translate.c:329:
Cannot write random bytes:
80524C50DB7F0000:error:4000008D:Luna gem engine:ERR_LUNACA3_error:PKCS#11 error:e_gem.c:1803:C_OpenSession=0x00000003
80524C50DB7F0000:error:4000008D:Luna gem engine:ERR_LUNACA3_error:PKCS#11 error:e_gem.c:1561:
```

**Reply by vyasatavi** [reply]
@Joelone Are those PKCS#11 logs cklogs? If not can you enable cklogs in Chrystoki.conf file?

**Comment by vyasatavi** [comment]
@mattcaswell Apparently Safenet tried this on OpenSSL 3.0.7 and did not see any error. Can you check if something introduced in 3.0.8 could have broken it?

**Comment by vyasatavi** [comment]
With help from Thales, we found a resolution for this issue. In OpenSSL 1.1.1, while using the Thales gem engine to use Thales Luna HSM,  we could pass a fake PEM encoded RSA key (ref key to actual key on HSM server)  to OpenSSL req or x509 command to generate a CSR or self-signed cert with -engine 'gem' and WITHOUT the use of -keyform option (engine or pkcs11). OpenSSL would not bother parsing it and just gives it to the engine.

Starting OpenSSL 3.0 I think it will attempt to parse it if -keyform is not provided. This was the issue. After providing -keyform engine the command worked without any issue. 

if there is no followup. I would like to close this discussion. 

CC @Joelone @mattcaswell
